Library/Homebrew/utils.rb

```diff
@@ -188,14 +188,32 @@ def self.homebrew_version_string
 
   def self.install_gem_setup_path!(gem, version = nil, executable = gem)
     require "rubygems"
-    ENV["PATH"] = "#{Gem.user_dir}/bin:#{ENV["PATH"]}"
 
-    args = [gem]
-    args << "-v" << version if version
+    if Gem::Specification.find_all_by_name(gem, version).empty?
+      ohai "Installing missing or updating outdated '#{gem}' gem"
+      install_args = %W[--no-ri --no-rdoc --user-install #{gem}]
+      install_args << "--version" << version if version
+
+      # Do `gem install [...]` without having to spawn a separate process or
+      # having to find the right `gem` binary for the running Ruby interpreter.
+      require "rubygems/commands/install_command"
+      install_cmd = Gem::Commands::InstallCommand.new
+      install_cmd.handle_options(install_args)
+      exit_code = 1 # Should not matter as `install_cmd.execute` always throws.
+      begin
+        install_cmd.execute
+      rescue Gem::SystemExitException => e
+        exit_code = e.exit_code
+      end
+      odie "Failed to install/update the '#{gem}' gem." if exit_code != 0
+    end
 
-    unless quiet_system "gem", "list", "--installed", *args
-      safe_system "gem", "install", "--no-ri", "--no-rdoc",
-                                    "--user-install", *args
+    # Make sure RubyGems `bin` directory is in `PATH` and is before all others.
+    path = ENV["PATH"].split(File::PATH_SEPARATOR)
+    unless path.first && path.first == "#{Gem.user_dir}/bin"
+      path.delete("#{Gem.user_dir}/bin")
+      path.unshift("#{Gem.user_dir}/bin")
+      ENV["PATH"] = path.join(File::PATH_SEPARATOR)
     end
```

Comment:
That's true, multiple instances don't create any problem I'm aware of. I only changed this because I didn't like the thought of having duplicates in there. I'm fine with going back to:

```ruby
ENV["PATH"] = "#{Gem.user_dir}/bin#{File::PATH_SEPARATOR}#{ENV["PATH"]}"
```